test/cpp/api/autograd.cpp

```diff
@@ -223,6 +225,39 @@ TEST(CustomAutogradTest, FunctionReturnsUndefined) {
     {MyFunction::apply(x)}, {x}, {}, false, false, true)[0].defined());
 }
 
+TEST(CustomAutogradTest, MaterializeGrads) {
+  struct MyFunction : public Function<MyFunction> {
+    static Variable forward(AutogradContext *ctx, Variable var) {
+      return var;
+    }
+
+    static variable_list backward(AutogradContext *ctx, variable_list grad_output) {  
+      return grad_output;
+    }
+  };
+
+  auto x = torch::ones(1, torch::requires_grad());
+  UndefinedGrad().apply({MyFunction::apply(x)})[0].backward();
+  ASSERT_VARIABLE_EQ(x.grad(), torch::zeros_like(x));
+}
+
+TEST(CustomAutogradTest, DontMaterializeGrads) {
+  struct MyFunction : public Function<MyFunction> {
+    static Variable forward(AutogradContext *ctx, Variable var) {
+      ctx->set_materialize_grads(false);
+      return var;
+    }
+
+    static variable_list backward(AutogradContext *ctx, variable_list grad_output) {  
+      return grad_output;
```

Comment:
nit: You can have the assert here inside the backward function to make sure the grad_output that was given to you match what you expect.